src/browser/pages/vscode.ts

```diff
@@ -54,6 +54,84 @@ export function getNlsConfiguration(document: Document) {
   return JSON.parse(nlsConfig) as NlsConfiguration
 }
 
+type RegisterRequireOnSelfType = {
+  // NOTE@jsjoeio
+  // We get the self type by looking at window.self.
+  self: Window & typeof globalThis
+  window: Window
+  nlsConfig: NlsConfiguration
+  options: Options
+}
+
+type RequireOnSelfType = {
+  baseUrl: string
+  recordStats: boolean
+  paths: {
+    [key: string]: string
+  }
+  "vs/nls": NlsConfiguration
+}
+
+/**
+ * A helper function to register the require on self.
+ *
+ * The require property is used by VSCode/code-server
+ * to load files.
+ *
+ * We extracted the logic into a function so that
+ * it's easier to test.
+ **/
+export function registerRequireOnSelf({ self, window, nlsConfig, options }: RegisterRequireOnSelfType) {
+  const errorMsgPrefix = "[vscode]"
+
+  if (!self) {
+    throw new Error(`${errorMsgPrefix} Could not register require on self. self is undefined.`)
+  }
+
+  if (!window) {
+    throw new Error(`${errorMsgPrefix} Could not register require on self. window is undefined.`)
+  }
+
+  if (!options || !options.csStaticBase) {
+    throw new Error(
+      `${errorMsgPrefix} Could not register require on self. options or options.csStaticBase is undefined or missing.`,
+    )
+  }
+
+  if (!nlsConfig) {
+    throw new Error(`${errorMsgPrefix} Could not register require on self. nlsConfig is undefined.`)
+  }
```

Comment:
These should be handled by the type checker right?

Reply:
For `nlsConfig` we already throw if unable to get it (`getNlsConfiguration`) and for `getOptions` we always return something even if getting the options fails so both those can be omitted.

It's not possible for `self` and `window` to be missing unless this is being loaded outside a browser context in which case nothing will work anyway haha

Reply:
Although those last two points are moot now with the latest changes! But I don't think we need to check that origin is undefined. If `window.location` exists `origin` will also exist so if somehow `window` is missing we will have already errored before getting here (`no such property location on window` or something like that).

Reply:
I would also argue that passing an undefined to a function that only accepts a string is an error on the caller's side so if we do really need to typecheck we could do it before calling the func. So for example:

```javascript
if (!window) {
  throw new Error()
}
registerRequireOnSelf(window)
```

Or, if we want to keep the typecheck inside the function we could change the types to `Window | undefined` to accurately reflect what the function accepts.

Reply:
> For nlsConfig we already throw if unable to get it (getNlsConfiguration) and for getOptions we always return something even if getting the options fails so both those can be omitted.

Ah true! Good point. I'll remove those.

Reply:
> Oh are these typechecks for unit tests that don't load a browser context?

Yes exactly!

> If so that makes sense to me, maybe we can move them out in that case to the calling side

Ah I didn't think of that. Almost like an integration test (if i can stretch the definition that far lol) but I see what you mean. Almost like defining a context where "X" is missing, we should `catch` before we call the function. Smart! I'll refactor.